api/src/DataPersister/Util/DataPersisterObservable.php

```diff
@@ -0,0 +1,151 @@
+        $this->onAfterUpdate = $noop;
+        $this->onBeforeRemove = $identity;
+        $this->onAfterRemove = $noop;
+    }
+
+    public function supports($data, array $context): bool {
+        return $this->dataPersister->supports($data, $context);
+    }
+
+    public function onBeforeCreate(Closure $onBeforeCreate): self {
+        $this->onBeforeCreate = $onBeforeCreate;
+
+        return $this;
+    }
+
+    public function onAfterCreate(Closure $onAfterCreate): self {
+        $this->onAfterCreate = $onAfterCreate;
+
+        return $this;
+    }
+
+    public function onBeforeUpdate(Closure $onBeforeUpdate): self {
+        $this->onBeforeUpdate = $onBeforeUpdate;
+
+        return $this;
+    }
+
+    public function onAfterUpdate(Closure $onAfterUpdate): self {
+        $this->onAfterUpdate = $onAfterUpdate;
+
+        return $this;
+    }
+
+    public function onBeforeRemove(Closure $onBeforeRemove): self {
+        $this->onBeforeRemove = $onBeforeRemove;
+
+        return $this;
+    }
+
+    public function onAfterRemove(Closure $onAfterRemove): self {
+        $this->onAfterRemove = $onAfterRemove;
+
+        return $this;
+    }
+
+    public function onCustomAction(CustomActionListener $customActionListener): self {
+        $this->customActionListeners[] = $customActionListener;
+
+        return $this;
+    }
+
+    public function onPropertyChange(PropertyChangeListener $propertyChangeListener): self {
+        $this->propertyChangedListeners[] = $propertyChangeListener;
+
+        return $this;
+    }
+
+    public function persist($data, array $context = []): object {
+        if ('post' === ($context['collection_operation_name'] ?? null)) {
+            $data = call_user_func($this->onBeforeCreate, $data);
+        } elseif ('patch' === ($context['item_operation_name'] ?? null)) {
+            $data = call_user_func($this->onBeforeUpdate, $data);
+        } else {
+            foreach ($this->customActionListeners as $listener) {
+                if (self::actionNameMatches($listener, $context)) {
+                    $data = call_user_func($listener->getBeforeAction(), $data);
+
+                    break;
+                }
+            }
+        }
+
+        $data = $this->dataPersister->persist($data, $context);
+
+        if ('post' === ($context['collection_operation_name'] ?? null)) {
+            call_user_func($this->onAfterCreate, $data);
+        } elseif ('patch' === ($context['item_operation_name'] ?? null)) {
+            call_user_func($this->onAfterUpdate, $data);
+        } else {
+            foreach ($this->customActionListeners as $listener) {
+                if (self::actionNameMatches($listener, $context)) {
+                    call_user_func($listener->getAfterAction(), $data);
+
+                    break;
+                }
+            }
+        }
+
+        $dataBefore = $this->requestStack->getCurrentRequest()->attributes->get('previous_data');
```

Comment:
Die PropertyChangeListeners hast du hinzugefügt weil du die `'previous_data'` in #1908 brauchen kannst, sehe ich das richtig? Ich bin noch nicht sicher ob das so ein üblicher Use Case wird dass ich gleich dieses ganze System dazubauen würde, und wenn doch, dann bin ich noch nicht sicher ob wir wirklich nur immer erst nach der Änderung reagieren wollen werden. Und von der Reihenfolge / Priorität der Listeners wissen wir bisher auch noch nicht, ob das mal relevant wird... Aber ich muss zugeben dass es schön wäre, wenn die Logik die nur bei Änderung einer Property läuft abgekapselt ist.